storage/src/mmr/mutable.rs

```diff
@@ -0,0 +1,512 @@
+    pub fn update(&mut self, hasher: &mut H, key: K, value: V) {
+        // Update the updates tree
+        let record_digest = Self::key_value_hash(hasher, &key, &value);
+        let tree_pos = self.updates.add(hasher, &record_digest);
+
+        // Set the is_active bit of the new record to 1.
+        self.is_active.append(hasher, true);
+
+        // Update the snapshot.
+        let new_location = self.records.len();
+        if let Some(old_location) = self.snapshot.insert(key.clone(), new_location) {
+            // Flip the is_active bit of the old record to 0.
+            let record = &self.records[old_location];
+            assert_eq!(record.key, key);
+            let bit_offset = leaf_pos_to_num(record.tree_pos);
+            assert!(self.is_active.get_bit(bit_offset));
+            self.is_active.set_bit(hasher, bit_offset, false);
+        }
+
+        // Append the new record to the records vector.
+        let new_record = UpdateRecord {
+            tree_pos,
+            key,
+            value,
+        };
+        self.records.push(new_record);
+        assert_eq!(self.records.len(), self.is_active.bit_count() as usize);
+    }
+
+    /// Delete `key` and its value from the store. Panic if `key` has no value.
+    pub fn delete(&mut self, hasher: &mut H, key: K) {
+        // Remove the key from the snapshot.
+        let old_location = match self.snapshot.remove(&key) {
+            Some(location) => location,
+            None => panic!("key has no value {}", key),
+        };
+
+        // Flip the is_active bit of the old record to 0.
+        let old_record = &self.records[old_location];
+        assert_eq!(old_record.key, key);
+        let bit_offset = leaf_pos_to_num(old_record.tree_pos);
+        assert!(self.is_active.get_bit(bit_offset));
+        self.is_active.set_bit(hasher, bit_offset, false);
+
+        // We include a new record for the deletion operation so that we can derive all of the
+        // store's state from the records structure alone.
+        let new_record = UpdateRecord {
+            tree_pos: Self::DELETED,
+            key,
+            value: old_record.value.clone(),
+        };
+        let record_digest = Self::key_value_hash(hasher, &new_record.key, &new_record.value);
+        self.records.push(new_record);
+        println!("pos!! {} {}", self.updates.size(), record_digest);
+
+        // Update the updates tree
+        self.updates.add(hasher, &record_digest);
+
+        // Set the is_active bit of the deletion record to 0.
+        self.is_active.append(hasher, false);
+        assert_eq!(self.records.len(), self.is_active.bit_count() as usize);
+    }
+
+    /// Return the root hash of the mutable MMR.
+    pub fn root(&self, hasher: &mut H) -> H::Digest {
+        let updates_root = self.updates.root(hasher);
+        let is_active_root = self.is_active.root(hasher);
+        let mut h = Hasher::new(hasher);
+        h.node_hash(0, &updates_root, &is_active_root)
```

Comment:
I suspect the 0 here is a placeholder for the "minimum" record (unpruned)?

Reply:
comment added